proposals/context-parameters.md

```diff
@@ -0,0 +1,560 @@
+* Context receivers are added to the scope and marked as such (so differently from the default receiver). They have lower priority than other receivers.
+* Named context parameters are added in the same way as value parameters, and marked as such. They have the same priority as regular value parameters.
+
+Note that context receivers are **not** added as default receivers. In particular, this means they are **not** accessible through the `this` reference.
+
+**§28** *(resolution, candidate sets)*: When **building candidate sets** during resolution, we consider context receivers part of the implicit scope, in a similar way as default receivers. However:
+
+* We consider only member callables which are marked with `context`.
+* When considering callables where the type of the context receiver appears as extension or dispatch receiver, we consider only those marked with `context`.
+
+Remember that the candidate set from context receivers has less priority than those from other receivers.
+
+```kotlin
+class A { context fun foo(x: Int) }
+class B { fun foo(x: Object) }
+
+// here 'foo' resolves to the one in 'B'
+// because 'B' is introduced in an inner scope
+// the fact that the one from 'A' is more specific doesn't play a role
+context(A) B.example1() = foo(5)
+
+// 'this' always resolves to 'B'
+context(A) B.example2() = this.foo(5)
+
+// to access the members from 'A' use a name, 'summon'
+context(A) B.example3a() = summon<A>().foo(5)
+context(a: A) B.example3b() = a.foo(5)
+context(a: A) B.example3b() = with(a) { foo(5) }
+```
+
+**§29** *(resolution, most specific candidate)*: When choosing the **most specific candidate** we follow the Kotlin specification, with one addition:
+
+* Candidates with declared context (even if empty) are considered more specific than those without a context.
+* In particular, a declared but empty context is not more specific than one with 1 or more context parameters.
+
+**§30** *(resolution, context resolution)*: Once the overload candidate is chosen, we **resolve** context parameters (if any). For each context parameter:
+
+* We traverse the tower of scopes looking for **exactly one** default receiver, context receivers, or named context parameter with a compatible type.
+* It is an ambiguity error if more than one value is found at the same level.
+* It is an overload resolution error if no applicable context parameter is found.
+
+We stress that each context parameter is resolved separately. There is no special "link" between all the context receivers and arguments introduced in the same declaration.
+
+```kotlin
+interface Logger {
+  context fun log(message: String)
+}
+class ConsoleLogger: Logger { ... }
+class FileLogger: Logger { ... }
+
+context(Logger) fun logWithTime(message: String) = ...
+
+context(ConsoleLogger, FileLogger) fun example1a() =
+  logWithTime("hello")  // ambiguity error
+  
+context(console: ConsoleLogger, file: FileLogger) fun example1b() =
+  logWithTime("hello")  // ambiguity error
```

Comment:
Wouldn't `logWithTime` be unresolved here instead of an ambiguity error? From my understanding, the type of a named context parameter isn't part of the scope inside the function. Or am I wrong here?

Reply:
No, in every case `logWithTime` is resolved to the top-level function. The difference is whether there's ambiguity or not in the "context resolution" phase.